### PR TITLE
CONSOLE-2376: Console plugins

### PR DESCRIPTION
**Pull request by jhadvig**
Console-operator related changes based on the [enhancement doc](https://github.com/openshift/enhancements/pull/441/files).
/assign @spadgett

### CODE REVIEW

**Comment by jhadvig** [comment]
@spadgett I've tested the PR with thew [API change](https://github.com/openshift/api/pull/764/files#diff-2d91c76ad45d9598ff5cf4b5a091f9641507a937859fca78eadcece3e6cae9e9).
I've put the plugins field into the `spec.customization.plugins`, but we can also put it directly on the `spec` (thought that would make a bit more sense to have it in the `customization` field).
After the field with and item is set on the operator config, the console-config ConfigMap is updated with it and the `console` pods are redeployed.

Updated operator config:
Screenshot 2020-11-05 at 11 55 10

Updated console-config CM:
Screenshot 2020-11-05 at 11 42 57


PTAL

**Comment by jhadvig** [comment]
just FYI the TC's are failing cause I've removed the hardcoded API changes in the `vendor/`

File: pkg/console/operator/sync_v400.go
Comment (vojtechszocs) [inline pkg/console/operator/sync_v400.go:338]: :+1: This means the ConfigMap will contain an intersection of enabled plugins (operator config) and available plugins (`ConsolePlugin` custom resources).

@spadgett This shouldn't interfere with enabling a dynamic plugin as part of operator install flow, right? Console UI can start loading a plugin only when its assets are deployed and available via Service, which means only after the operator install process completes.

@jhadvig Will the ConfigMap be updated upon a change of either operator config or `ConsolePlugin` custom resources?

So if I understand correctly, operator config tracks all plugins marked as enabled (covers operator install flow and plugin admin page) and the ConfigMap tracks enabled plugins which should be loaded in the UI.

File: test/e2e/plugins_test.go
Comment (vojtechszocs) [inline test/e2e/plugins_test.go:57]: Please update this field to `BasePath` according to https://github.com/openshift/api/pull/764#discussion_r533645627

**Review by vojtechszocs — CHANGES_REQUESTED** [review]

File: pkg/console/operator/sync_v400.go
Comment (jhadvig) [reply pkg/console/operator/sync_v400.go:338]: @vojtechszocs 
> This shouldn't interfere with enabling a dynamic plugin as part of operator install flow, right?

No it shouldn't. It's an mechanism that prevents console from trying to load plugins that are not present on the cluster as a `ConsolePlugin` CR

> Will the ConfigMap be updated upon a change of either operator config or ConsolePlugin custom resources?

Yes. Both changes should redeploy the console pods

>  So if I understand correctly, operator config tracks all plugins marked as enabled (covers operator install flow and plugin admin page) and the ConfigMap tracks enabled plugins which should be loaded in the UI.

Yes. operator config tracks all the enabled plugins, but since we can not guarantee that are of these plugins are available on the cluster (since the operator config can be updated manually as well) operator will check if a corresponding `ConsolePlugin` exists on the cluster. If so then it will put int into the `console-config.yaml`.

File: pkg/console/operator/operator.go
Comment (stlaz) [inline pkg/console/operator/operator.go:150]: are you sure? The events from these informers are filtered so that only changes to resources of name `api.OpenShiftConsoleName` are caught.

The way I understand the API is that there are many objects of the `ConsolePlugin` type that you care about, but I might be wrong.

File: pkg/console/operator/operator.go
Comment (stlaz) [inline pkg/console/operator/operator.go:121]: you only appear to be doing a `List()` over console plugins, you should use a lister (it caches the data so that you don't need to do live requests to the API server)

File: pkg/console/operator/sync_v400.go
Comment (stlaz) [inline pkg/console/operator/sync_v400.go:517]: use the power of sets instead of having to write this on your own https://github.com/kubernetes/kubernetes/blob/7d72ddca6e9a4e9cfa481b137bd0546be6a16cad/vendor/k8s.io/apimachinery/pkg/util/sets/string.go#L30

**Review by stlaz — CHANGES_REQUESTED** [review]
QoL changes, both for the API server and the developer

File: pkg/console/operator/operator.go
Comment (jhadvig) [reply pkg/console/operator/operator.go:151]: @stlaz right, so we dont need to register the `consolePluginInformer` here ?

**Comment by jhadvig** [comment]
@spadgett the PR is ready to review.
The logic is following:
When admin adds a list of ConsolePlugins into the console-operator config's `spec.plugins` field, only ConsolePlugins that are available on the cluster, meaning those that are not wont get saved. After that console-operator will fetch all the available and enabled ConsolePlugins CRs and creates a `map[string]string`, where the key is the name of the ConsolePlugin CR and the value is the endpoint to which the request from console will be proxied. The map is passed into the `console-config.yaml`.
I know that originally we wanted to pass only the list of enabled plugins into the `console-config.yaml`, but during development that I've realised that it would be easier just to assemble the endpoints by the console-operator and just pass them to the console in the config, cause:
- it removes duplicate API calls for fetching the ConsolePlugin CRs (console-operator when checking if the CRs are available and console itself when assembling the endpoints)
- it will  make local testing easier for the devs that will rewrite the static plugin into a dynamic one, since they can start the bridge with passing the `--plugins <pluginName>=<pluginEndpoint>` flag and point it to a locally running server on their host for serving assets
PTAL

File: pkg/console/operator/operator.go
Comment (stlaz) [inline pkg/console/operator/operator.go:147]: `.WithInformers(consolePluginInformer)` to sync on console plugin events?

File: pkg/console/operator/sync_v400.go
Comment (stlaz) [inline pkg/console/operator/sync_v400.go:525]: this never gets called?

**Review by stlaz — CHANGES_REQUESTED** [review]
looks like there's some dead code and the controller won't sync on console plugin events, otherwise looks good

**Comment by jhadvig** [comment]
@stlaz thanks for review, comments addressed.

/hold for approvals
/assign @ahardin-rh @yapei @sferich888

**Comment by yapei** [comment]
LGTM

File: pkg/console/operator/sync_v400.go
Comment (stlaz) [inline pkg/console/operator/sync_v400.go:513]: you may want to consider going "Progressing: true" in this case

File: pkg/console/subresource/consoleserver/config_builder.go
Comment (stlaz) [inline pkg/console/subresource/consoleserver/config_builder.go:282]: why do you avoid returning the "initialized, but no keys" path?

**Comment by stlaz** [comment]
/lgtm
/hold
from what I've been told, you don't merge API before all related PRs get lgtm, so /holding for API merge

**Review by spadgett — APPROVED** [review]
/lgtm
/hold cancel

**Comment by spadgett** [comment]
/hold

We probably need to delete the CI project or wait a few hours or this will keep failing. The record set didn't get cleaned up from a previous run.

```
level=error msg=Error: [ERR]: Error building changeset: InvalidChangeBatch: [Tried to create resource record set [name='api.ci-op-qlxhpr6t-4e531.origin-ci-int-aws.dev.rhcloud.com.', type='A'] but it already exists]
level=error msg=	status code: 400, request id: ac72bd30-3bee-4862-b034-4e30be6edc4c
level=error
level=error msg=  on ../tmp/openshift-install-882015745/route53/base.tf line 36, in resource "aws_route53_record" "api_external_alias":
level=error msg=  36: resource "aws_route53_record" "api_external_alias" {
level=error
level=error
level=fatal msg=failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed to apply Terraform: failed to complete the change
error: failed to execute wrapped command: exit status 1 
```

**Pull request merged**
